h/views/account_signup.py

```diff
@@ -76,15 +81,172 @@ def validation_failure(self):
+        )
+
+    @exception_view_config(
+        context=IDInfoJWTDecodeError, renderer="h:templates/error.html.jinja2"
+    )
+    def idinfo_jwt_decode_error(self):
+        report_exception(self.context)
+        self.request.response.status_int = 403
+        return {"error": "Decoding idinfo JWT failed."}
+
+    @exception_view_config(context=ValidationFailure)
+    def validation_failure(self):
+        orcid_id = self.decode_orcid_id()
+        self.request.response.status_int = 400
+
+        # When reloading the page prefill the form with the previously submitted values.
+        form_data = {
+            "username": self.request.POST.get("username", ""),
+            "privacy_accepted": self.request.POST.get("privacy_accepted", "") == "true",
+            "comms_opt_in": self.request.POST.get("comms_opt_in", "") == "true",
+        }
+
+        # When reloading the page replace the idinfo token with a fresh one.
+        form_data.update(
+            encode_idinfo_token(
+                self.jwt_service, orcid_id, JWTIssuers.SIGNUP_VALIDATION_FAILURE_ORCID
+            )
+        )
+
+        return {
+            "js_config": {
+                "formErrors": self.context.error.asdict(),
+                "formData": form_data,
+                **self.js_config(orcid_id),
+            }
+        }
+
+    def js_config(self, orcid_id):
+        feature_service = self.request.find_service(name="feature")
+
+        return {
+            "csrfToken": get_csrf_token(self.request),
+            "features": {
+                "log_in_with_orcid": feature_service.enabled(
+                    "log_in_with_orcid", user=None
+                )
+            },
+            "identity": {"orcid": {"id": orcid_id}},
+        }
+
+    def decode_orcid_id(self):
+        try:
+            idinfo = self.jwt_service.decode_symmetric(
+                self.request.params["idinfo"],
+                audience=JWTAudiences.SIGNUP_ORCID,
+                payload_class=IDInfo,
+            )
+        except JWTDecodeError as err:
+            raise IDInfoJWTDecodeError from err
+
+        return idinfo.sub
 
 
 # It's possible to try to sign up while already logged in. For example: start
 # to signup but don't submit the final form, then open a new tab and log in,
 # then return to the first tab and submit the signup form. This view is called
 # in these cases.
 @view_config(route_name="signup", is_authenticated=True)
+@view_config(route_name="signup.orcid", is_authenticated=True)
 def is_authenticated(request):
     return HTTPFound(
         request.route_url("activity.user_search", username=request.user.username)
     )
+
+
+def encode_idinfo_token(jwt_service, orcid_id, issuer):
+    return {
+        "idinfo": jwt_service.encode_symmetric(
+            IDInfo(orcid_id),
+            expires_in=timedelta(hours=1),
+            issuer=issuer,
+            audience=JWTAudiences.SIGNUP_ORCID,
+        ),
+    }
```

Comment:
This is pulled out as a separate helper function because there are multiple places (multiple different "issuers") that need to encode these idinfo tokens for the `/signup/orcid` page:

1. The `/oidc/redirect/orcid` view (in `views/oidc.py` below) encodes one when redirecting to the `/signup/orcid` view
2. When invalid form fields are submitted and the `/signup/orcid` view responds by re-rendering the form it encodes a new idinfo token for itself

We use two different `iss` strings for these two different places in the code where idinfo tokens are encoded. This could be useful if inspecting token contents when debugging. If necessary in future the code consuming these tokens could vary its behaviour depending on the issuer.